Remove the incomplete-downloaded cache if interrupted while pulling

WestleyK opened the pull request:
## Description of the Pull Request (PR):

This PR fixes the issue when you pull a image, then interrupt the pull, then pull it again.

## This fixes or addresses the following GitHub issues:

- Fixes #3788 
- Fixes #3776 

<br>

Commit:
Remove the half-downloaded cache if interupted while pulling

Commit:
Removed commented code

Commit:
Fixed spelling + cleaned code + added interupt check for library

Commit:
Removed commented code

ikaneshiro commented on internal/pkg/client/cache/library.go:
```diff
@@ -53,7 +55,7 @@ func (c *Handle) LibraryImageExists(sum, name string) (bool, error) {
 	}
 	if cacheSum != sum {
-		return false, fmt.Errorf("cached file sum (%s) and expected sum (%s) does not match", cacheSum, sum)
+		return false, ErrCacheHashNotTheSame
```
Can we do the same in `oras.go` if it's returning an error for this case too

ikaneshiro commented on internal/pkg/client/cache/library.go:
```diff
@@ -6,13 +6,15 @@
 	"github.com/sylabs/scs-library-client/client"
 )
 
+var ErrCacheHashNotTheSame = errors.New("hash does not match")
```
Maybe we should put this in `dir.go` since this error will now be part of the package API. Might also be worth calling it something like `ErrBadChecksum` so it's a little less verbose.

ikaneshiro commented on internal/app/singularity/pull.go:
```diff
@@ -311,6 +351,17 @@ func OciPull(imgCache *cache.Handle, name, imageURI, tmpDir string, ociAuth *oci
+		// monitor for OS signals and remove invalid file
+		c := make(chan os.Signal)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+		go func(fileName string) {
```
Maybe we should start considering a general signal handler implementation. This is very similar to the signal handling in `build.go` and we can probably create a package to do what we want. I wouldnt let this block the PR, but make a tech debt issue if you dont want to address it in this PR

Commit:
Moved signal-interrupt to a common function +
ErrHashNotTheSame to ErrBadChecksum & moved it to dir.go

Commit:
Added the "if err = cache.ErrBadChecksum" to oras

Commit:
Added godoc comment for SignalHandlerInterrupt & updated error check for that function

Commit:
Removed extra newline, to be consistent
I'm also trying to use less commits, but thats not working out very whell. Theres always something to change :/

Commit:
Changed function name from "SignalHandlerInterrupt" to "InterruptCleanup" as IanK suggested

gvallee commented on internal/app/singularity/interrupt_cleanup.go:
```diff
@@ -0,0 +1,30 @@
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+)
+
+// InterruptCleanup will watch for a interrupt signal, if theres
```
```suggestion
// InterruptCleanup will watch for a interrupt signal, if there's
```

mem replied:
please don't export unless you need it in a different package.

gvallee reviewed (approved):
I just found a typo, otherwise LGTM. Thanks for implementing that.

mem commented on internal/app/singularity/interrupt_cleanup.go:
```diff
@@ -0,0 +1,30 @@
+// InterruptCleanup will watch for a interrupt signal, if theres
+// one detected, then it will remove all the specified file(s)
+func InterruptCleanup(files ...string) {
+	c := make(chan os.Signal)
```
the channel used to receive signal notifications should be buffered. A buffer size of 1 is enough in this case.

mem commented on internal/app/singularity/pull.go:
```diff
@@ -68,11 +68,19 @@ func LibraryPull(imgCache *cache.Handle, name, ref, transport, fullURI, libraryU
+		go InterruptCleanup(imagePath)
 
 		// call library download image helper
 		if err = library.DownloadImage(context.TODO(), libraryClient, imagePath, imageRef, downloadImageCallback); err != nil {
```
since this is using a context, the signal handler should use it to notify the download process that it should stop. I would expect that if the download is stopped, this functions returns an error, and you can let the normal code flow continue from here (dealing with the error).

This also hints at the signal handler being installed "too low". It should probably be installed higher in the call stack so that you can pass a context down and handle cancellation in that way, as it is likely that all clients need the same handling.

GodloveD replied:
This is a good idea, but I think it should be added as an additional improvement (see #3899).  Since this is fixing a bug, I think it can be added as it already is, and the context can be fixed later.

Commit:
Unexported InterruptCleanup(), suggested by @mem +
fixed type suggested by @gvallee (theres => there's)

Commit:
Added buffer to interruptCleanup function, as mem suggested

Commit:
Merge remote-tracking branch 'upstream/master' into shub_cache_fix

mem reviewed (commented):
Removing my request for changes, but not approving.

Commit:
Merge remote-tracking branch 'upstream/master' into shub_cache_fix
